op-program/chainconfig/chaincfg.go

```diff
@@ -19,6 +26,37 @@ func init() {
 	}
 	OPSepoliaChainConfig = mustLoadConfig(11155420)
 	OPMainnetChainConfig = mustLoadConfig(10)
+
+	// Load custom l2 genesis and rollup config from embed FS
+	data, err := customChainConfigFS.ReadFile("configs/genesis-l2.json")
+	if err != nil {
+		panic(err)
+	}
+	var genesis core.Genesis
+	err = json.Unmarshal(data, &genesis)
+	if err != nil {
+		panic(err)
+	}
+	CustomChainConfig = genesis.Config
+
+	file, err := customChainConfigFS.Open("configs/rollup.json")
+	if err != nil {
+		panic(err)
+	}
+	dec := json.NewDecoder(file)
+	dec.DisallowUnknownFields()
+	if err := dec.Decode(&CustomRollupConfig); err != nil {
+		panic(err)
+	}
+	// Both configs must have the same L2 chainID
+	if CustomChainConfig.ChainID.Uint64() != CustomRollupConfig.L2ChainID.Uint64() {
+		panic(fmt.Errorf("mismatched genesis-l2.json chainid %d vs rollup.json chainid %d", CustomChainConfig.ChainID.Uint64(), CustomRollupConfig.L2ChainID.Uint64()))
+	}
+	// Do not override existing superchain registered configs
+	if _, err := params.LoadOPStackChainConfig(CustomChainConfig.ChainID.Uint64()); err == nil {
+		panic(fmt.Errorf("cannot override existing superchain registered config"))
+	}
+	L2ChainConfigsByChainID[CustomChainConfig.ChainID.Uint64()] = CustomChainConfig
```

Comment:
Note: nothing actually reads from this map - it's removed in https://github.com/ethereum-optimism/optimism/pull/12307